[RFC] Foundation for fulltext filtering

bstoeger opened the pull request:
<!-- Lines like this one are comments and will not be shown in the final output. -->
<!-- Make sure that you have read the "Contributing" section of the README and also the notes in CodingStyle. -->
<!-- If you are a collaborator, please add labels and assign other collaborators for a review. -->

### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [x] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
<!-- Describe your pull request in detail. -->
This is a primitive first implementation of a fulltext index. There are two reasons for this:
1) The filterwidget on desktop is getting horrendously complex and I wonder if many usecases couldn't be replaced by a simple fulltext search. My suggestion would be to show only the fulltext search first and then let the user expand "advanced options".
2) IIRC we do fulltext search on mobile in a background thread, which appears to be very brittle. Therefore, I suggest using the same fulltext code on mobile and desktop.

Be warned that this is at concept stage, though it worked in my preliminary tests. It is 100% unoptimized and there is ample room for optimization. However, when testing on a ~4000 dives testlog, the actual fulltext search was always faster than 10 ms. If the search-words reach three letters or more, it typically got even <1 ms. I wonder if optimizing the actual search is even worth it.

There is one big downside: currently, loading the testlog mentioned above is slowed by ~750 ms, because the index is generated on loading. A first attempt at generating the index in a background thread failed: The profile widget clears the `displayed_dive`, which accesses the cache and thus waits for the index generation. Moreover, the background thread would somehow have to lock on the dive-list, which we don't even support. Ugh.

If someone suggests an external fulltext-library that is easy to hook into, I'm fine with that too!

### Additional information:
<!-- Include sample dive log or other relevant information to allow testing the change where feasible. -->

### Release note:
<!-- Describe if this change needs a release note present in CHANGELOG.md. -->
<!-- Also, please make sure to add the release note on top of the file CHANGELOG.md. -->
Let's stabilize this first.
### Documentation change:
<!-- If this PR makes changes to user functionality, then the documentation has to be updated too. -->
<!-- Please, briefly outline here what has changed in terms of the user experience (UX). -->
<!-- If UX changes have been made, a maintainer should apply the 'needs-documentation-change' label. -->
Let's stabilize this first.
### Mentions:
<!-- Mention users that you want to review your pull request with @<user-name>. Leave empty if not sure. -->
@dirkhh @willemferguson @atdotde

dirkhh commented:
I read through the code. What would you like feedback / testing on?

bstoeger commented:
Typing on phone, sorry for brevity:

1) Good idea? Use existing (which?) library instead?

2) What about slower startup/loading of huge divelogs? We fundamentally have no locking of the dive list and implementig it would be a huge project. Perhaps offset by improving the rest of the startup time? Or try to make indexing faster?

3) Testing would be basically the same as for the other filter PR: does it work and how is the interaction with the undo machinery?

Thanks!

dirkhh commented:
1. I wouldn't use an existing library as that likely would be way too generic and add overhead, dependencies, pain of compilation on different platforms, etc. Whether or not this is a good idea... I didn't feel the urgent need for it (😄) but it doesn't seem to hurt, either.
2. The question is "what's the real world start time on older phones". That's what needs benchmarking - a few tenths of a second are fine. Several seconds would be a show stopper
3. This currently is only on mobile, where in master we don't have undo. So I'm a bit confused what you want me to test here

bstoeger commented:
> This currently is only on mobile, where in master we don't have undo. So I'm a bit confused what you want me to test here

Sorry - miscommunication. This is currently only on desktop and I plan to unify with mobile later, when we implement filter on the "new" mobile dive list. My fear is that undo and searching in background can lead to disastrous results. If a search takes a few seconds, the user has enough time to undo an old command and boom - unprotected concurrent access.

dirkhh commented:
with the large anonymous test data file (8058 dives, 206MB xml data file) the startup time is __SIGNIFICANT__ - but filter time is perfectly reasonable:
~10s for loading the file, 273ms of that for the filter reload
admittedly, tested on a very fast system -- then again, this is a fairly extreme data file :-)

interactive performance when using the filter is not instantaneous, but perfectly reasonable.

From a performance perspective I don't see a reason not to merge this.

bstoeger commented:
Update:
- Move fulltext field to top of filterwidget.
- Focus fulltext field when widget is shown.
- Consider trip- and divesite-names in fulltext search.
- Add CHANGELOG.md entry.

Notes:
- The tab-order in the filterwidget is completely ridiculous!
- The UI is now broken with respect to column widths.
- The fulltext does not search for trip- or divesite-notes. I wonder how we should treat that.

dirkhh commented:
Tested some more and I like it.
I would of course love to hear from some others...

bstoeger commented:
Arghl. Looking at the mobile filter I just realized that I had a few misconceptions:
- The mobile code does not filter in the background owing to performance, but because of some concurrent access reasons. The code looks very scary!
- The mobile filter has the option of case-sensitivity. We could expand the index to save both by case-insensitive and case-sensitive. Or, we could just use the index for a case-insensitive lookup and then do the case-sensitive check on top.
- Likewise, the including notes vs. not-including notes feature. I will have to ponder on that.

Seems like reusability is less than I hoped. :-(

dirkhh commented:
Or, alternatively, we could harmonize semantics.
No reason to do case sensitive searches, frankly.

Excluding the notes was added because we don't have searches for just people or just tags. This seemed like a reasonable compromise.

I'm open to other ideas

bstoeger commented:
> Excluding the notes was added because we don't have searches for just people or just tags. This seemed like a reasonable compromise.

I added people and tag search modes on mobile and let the fulltext search use the fulltext index. I experienced random crashes that seemed timing-related (crashes in different threads, not reproducible in valgrind). Therefore, I removed the background-thread thing and since then no crash (knock wood).

Please let me know if this is - in principle - acceptable for you. As you would expect, my QML is horrible.

The advantage of doing this here instead of our divelist branch is that it allows for less divergence between both branches. Of course, it will make the next merge harder, yes.

dirkhh commented:
> > Excluding the notes was added because we don't have searches for just people or just tags. This seemed like a reasonable compromise.
> 
> I added people and tag search modes on mobile and let the fulltext search use the fulltext index. I experienced random crashes that seemed timing-related (crashes in different threads, not reproducible in valgrind). Therefore, I removed the background-thread thing and since then no crash (knock wood).
> 
> Please let me know if this is - in principle - acceptable for you. As you would expect, my QML is horrible.
> 
> The advantage of doing this here instead of our divelist branch is that it allows for less divergence between both branches. Of course, it will make the next merge harder, yes.

I'm certainly ok with removing the concurrency (we'll see if that causes a significant performance issue somewhere). I seem to remember that this was first introduced because of an awkward UI hang... but the fact that there is no locking of data structures is of course, err, regrettable.

I'll deal with the merge nightmares when they come up. Forward progress is my main goal.
In some small way I wonder if it wouldn't be better to focus on getting our branch ready to be merged before working on other things, but that's of course easier said than done.

bstoeger commented:
> In some small way I wonder if it wouldn't be better to focus on getting our branch ready to be merged before working on other things, but that's of course easier said than done.

This commit is part of it, as it unifies filtering. Of course, this probably could have been solved in a more pragmatic / less radical way. :-P

dirkhh commented:
> > In some small way I wonder if it wouldn't be better to focus on getting our branch ready to be merged before working on other things, but that's of course easier said than done.
> 
> This commit is part of it, as it unifies filtering. Of course, this probably could have been solved in a more pragmatic / less radical way. :-P

Our other branch is based on master as of a couple of days ago. Should this be merged into master before we merge our branch?

bstoeger commented:
> Should this be merged into master before we merge our branch?

Yes, but no hurry. Let's make sure that it is stable first. Those random crashes made me nervous, though I didn't see one after removing the multi-threading thing.

Once this is merged, I'll do the rebase of our branch.

bstoeger commented on desktop-widgets/filterwidget2.ui:
```diff
@@ -35,109 +35,95 @@
-         <property name="text">
-          <string>Tags</string>
-         </property>
+       <item row="13" column="2">
```
Off-topic: I have been ridiculed for creating my UIs with normal C++ function calls instead of these XML-files. At least it gives reasonable and mergeable diffs, compared to this mess!

dirkhh commented:
I have merged the underlying #2610, but not this one. Have you figured out what caused the crashes? I played with it again and didn't see crashes and didn't perceive the additional slowdown at load to be a major issue...

bstoeger commented:
> I have merged the underlying #2610, but not this one.

PR rebased cleanly.

> Have you figured out what caused the crashes?

Not seen crashes since I removed the QtConcurrent thing. That said, I haven't tested it extensively either! Crashes were from different thread contexts and I couldn't hit them with valgrind, which suggests some concurrent access thing. Though no proof, obviously.

dirkhh commented:
Testing on my slowest phone with 8000 dives / 170MB dive file. Startup is... slow. just under 20 seconds, 3.5 seconds until we see the empty dive list, another 16 for the divelist to populate.
We'd definitely need some form of 'waiting' thing there. Doesn't have to be an animated progress bar, but at least an indication that something is happening.
On a small screen the drop down for Fulltext/People/Tags is taking up way too much space, but that's fairly easy to fix.
What does bother me is that the number of matching dives in that search bar is no longer updated. That's definitely a bug.

dirkhh commented:
I am being rude again, pushing into your branch. This fixes the styling issue with the filterHeader for me. I still haven't figured out why the `diveModel.shown()` call no longer gives us the number of filtered dives... I'm suspecting it's using the number of rows of the wrong model...

dirkhh commented:
After some more testing I am mostly ok with this. Typo in commit message, issue with number of filtered dives, other than that this should be fine. And of course the DiveList.qml changes will have to be merged into the rewritten dive list in our other branch as part of the bigger rebase (which will be a pain).
I know you are traveling, I'll give this some time :-)

bstoeger commented:
> I am being rude again, pushing into your branch. This fixes the styling issue with the filterHeader for me.

Thanks, I squashed one of the commits. Also removed the "myobile" typo.

>  I still haven't figured out why the diveModel.shown() call no longer gives us the number of filtered dives... 

The call is not called anymore because we don't disconnect/connect the model since removing the multi-threading thing. I tried it with a Q_PROPERTY, please check it out. (For obvious reasons I don't want to invest too much time in the old dive list!).

> And of course the DiveList.qml changes will have to be merged into the rewritten dive list in our other branch as part of the bigger rebase (which will be a pain).

I can give it a try.

> I know you are traveling, I'll give this some time :-)

Indeed, no more work from me today.

PS: Mac build breaks because of kirigami-something...!?

bstoeger commented:
Help? All mobile builds are failing because of missing files?

dirkhh commented:
I'm traveling as well, will look this evening

Subsurface-divelog commented:
breeze deleted a ton of icons. Easiest fix is to go back to yesterdays version (well, they now use scripts to create them - but that really doesn't seem worth it for us...)
I'll push a 'fix' to master which you could rebase on, if you want

dirkhh commented:
WARNING!!!
=========

Force pushed the branch to rebase on latest master - this should fix the build issues.

dirkhh commented:
OK, I did end up adding another commit to do a better job with the drop down. And yes, QML Layouts are very odd beasts.

willemferguson commented:
On 2020/02/24 15:14, Dirk Hohndel wrote:
>
> OK, I did end up adding another commit to do a better job with the 
> drop down. And yes, QML Layouts are very odd beasts.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Subsurface-divelog/subsurface/pull/2625?email_source=notifications&email_token=ABGX5MS5RA3GVIZBNWXNLJTREPBZ3A5CNFSM4KWHNHM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMXXSOY#issuecomment-590313787>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/ABGX5MUJN5BMHVZD3UZWGADREPBZ3ANCNFSM4KWHNHMQ>.
>
I built and ran this branch. Here are some comments as a user, not as a 
programmer.

The filter fields has 3 options:

1. starts with

2. exact

3. substring

It is not clear what these options mean. When I select exact and search 
for "scubapro drysuit", I get dives that also contain in the suit text 
box:  "Scubapro undersuit, Argonaut drysuit". Clearly this is not what I 
am looking for. This happens irrespective of which of the above three 
options I selected.

It appears that some of the implementation details will be finalised a 
bit later and that the current focus is on creating the data structure.

My log as some 600 dives and I did not perceive any delay in loading the 
dive log. Neither did I encounter any abnormal behaviour or segfaults.

Kind regards,

willem




-- 
This message and attachments are subject to a disclaimer.

Please refer to 
http://upnet.up.ac.za/services/it/documentation/docs/004167.pdf 
<http://upnet.up.ac.za/services/it/documentation/docs/004167.pdf> for
full 
details.

dirkhh commented:
> I built and ran this branch. Here are some comments as a user, not as a programmer. 
> The filter fields has 3 options:
> 1. starts with
> 2. exact
> 3. substring
>
> It is not clear what these options mean. When I select exact and search for "scubapro drysuit", I get dives that also contain in the suit text box:  "Scubapro undersuit, Argonaut drysuit". Clearly this is not what I am looking for. This happens irrespective of which of the above three options I selected. 
>
> My log as some 600 dives and I did not perceive any delay in loading the dive log. Neither did I encounter any abnormal behaviour or segfaults.

Thanks for testing @willemferguson - this helps. @bstoeger is traveling for work, I'll wait for him to comment, but it's good news to see that you see neither performance issues nor crashes.

bstoeger commented:
This works on a word-by-word basis (match-all) and the mode (starts-with, exact, substring) only affects these single-word matches (as we currently do in the tag and people filters). We could, of course, add whole expression searches, either by adding a different mode (e.g. "phrase"?), or by recognizing quotation marks as Google et al. do. However, I presume that currently @dirkhh would not be too happy about investing a significant amount of time into that.

If the options as they are now turn out to be too confusing, I suggest removing the substring option and rename the others (starts-with and exact) to something more appropriate.

dirkhh commented:
so 'exact' would be maybe better called 'full word'?
What @willemferguson is looking for "kind of" makes sense especially for tags where 'exact' indeed could be thought to mean "that's exactly the tag". But what you are doing here is really just checking that it is these words and that we aren't matching a substring of a word. cave/cavern

I think what we have here is ready to go in, TBH

bstoeger commented:
> so 'exact' would be maybe better called 'full word'?

Yes.

> What @willemferguson is looking for "kind of" makes sense especially for tags where 'exact' indeed could be thought to mean "that's exactly the tag"

I just realized that we do support tags with whitespace in them. I will think about how to treat them with respect to fulltext search.

bstoeger commented:
Squashed commit. Changed "Exact" to "Full word" for fulltext-filter.

For @willemferguson's use case, people can still use the tags-filter until we implement a "full phrase" search.

dirkhh commented:
I am completely at a loss what's going on with the Android build...???

bstoeger commented:
The strange thing is that #2638 doesn't exhibit the Android problem. Do I have to completely remake the branch here...? Can you reproduce that locally?

dirkhh commented:
The error seems completely bogus. I keep hoping that restarting the build will help, but no luck.
And I can build this locally for 32 and 64 bit.

dirkhh commented:
merged via #2641
